README
ADDITIONAL INFORMATION
----------------------

S25 team needs to support Symfony 1.4 for legacy projects so
we've added a support PHP 7.4 and PHP 8.0 to Symfony 1.4 framework.

System requirements
-------------------

1. PHP 7.4 or PHP 8.0

Support's here
--------------

https://github.com/shop25/symfony

Old Description
---------------

Symfony is a complete framework designed to optimize the development of web applications by way of several key features.
For starters, it separates a web application's business rules, server logic, and presentation views.
It contains numerous tools and classes aimed at shortening the development time of a complex web application.
Additionally, it automates common tasks so that the developer can focus entirely on the specifics of an application.
The end result of these advantages means there is no need to reinvent the wheel every time a new web application is built!

Symfony was written entirely in PHP 5.
It has been thoroughly tested in various real-world projects, and is actually in use for high-demand e-business websites.
It is compatible with most of the available databases engines, including MySQL, PostgreSQL, Oracle, and Microsoft SQL Server.
It runs on *nix and Windows platforms.

LICENSE
-------

see LICENSE file

INSTALLATION
------------

see http://www.symfony-project.org/getting-started/1_4/en/


CHANGELOG
CHANGELOG
=========

12/28/20: Version 1.4.21
------------------------
* added support of PHP 7.4 and PHP 8.4

11/25/12: Versions 1.4.20
-------------------------

 * [33598] fixed the possibility to fake a file upload
 * [33545] fixed sfPDOSessionStorage for Oracle (closes #10022)
 * [33544] fixed sfWebRequest::splitHttpAcceptHeader incorrect result order (closes #10069, patch by Keri Henare)
 * [33539] fixed exception format when using the PHP 5.4 built-in server (closes #10067, based on a patch from jgskin)
 * [33486] fixed sfPDODatabase::call() method (closes #10044)

10/09/12: Versions 1.4.19
-------------------------

 * [33545] fixed sfPDOSessionStorage for Oracle (closes #10022)
 * [33544] fixed sfWebRequest::splitHttpAcceptHeader incorrect result order (closes #10069, patch by Keri Henare)
 * [33539] fixed exception format when using the PHP 5.4 built-in server (closes #10067, based on a patch from jgskin)
 * [33486] fixed sfPDODatabase::call() method (closes #10044)

05/30/12: Versions 1.4.18
-------------------------

 * [33466] fixed a possible DB session fixation attack (patch from Dmitri Groutso)
 * [33373] fixed test browser click function does not handle css selector without [ or ] (closes #9982, patch from mouette)

03/08/12: Versions 1.4.17
-------------------------

 * [33363] added some tests (closes #3237, patch from Stephen.Ostrow)
 * [33362] fixed a notice on PHP 5.4 (closes #9985, patch from bshaffer)
 * [33358] fixed Notice with PHP 5.4 (closes #10003)
 * [33309] fixed error in Debug mode from sfDebug.class.php when myUser implements sfSecurityUser (closes #9996)
 * [33299] reverted [33226] because of side effects (refs #8348)
 * [33292] fixed test for PHP 5.3 (patch from pylebecq)

12/13/11: Version 1.4.16
------------------------

 * [33251] fixed sfChoiceFormat when a string to translate contains a valid range (closes #9973)
 * [33250] fixed saving i18n fields in subforms (closes #7626, patch from yoshy71)
 * [33249] fixed last modified date calculation in sfMessageSource_Aggregate (closes #9981, patch from jamiel)
 * [33226] fixed merging problem for the routing configuration (closes #8348)
 * [33214] fixed ob_start() behavior on CLI (closes #9970)
 * [33208] fixed ob_start usage (to avoid warning in PHP 5.4, closes #9970)

10/27/11: Version 1.4.15
------------------------

 * [33151] fixed usage of mb_strlen in tasks (closes #9940)
 * [33149] added missing admin.delete_object event (closes #9943)
 * [33137] fixed multiple database support in Propel plugin (for the generator and sfPropelData, closes #8345)
 * [33125] fixed the possibility to include files included in an exclude rule in the deploy task (closes #9912)
 * [33122] fixed include|get_component when sfPartialView class is customized (closes #9932)
 * [33121] fixed protocol relative URL in the asset helper (closes #9936)
 * [33053] fixed typo (closes #9927, based on a patch from pmallet)

09/16/11: Version 1.4.14
------------------------

 * [33025] fixed sfCacheSessionStorage does not always check whether HTTP_USER_AGENT is set (closes #9921, patch from boutell)
 * [33022] fixed auto_link_text() when an email address is already linked (closes #9915, based on a patch from klemens_u)
 * [33021] fixed sfCacheSessionStorage does not serialize properly, also uses wrong option name for httponly (closes #9919, based on a patch from boutell)
 * [32939] fixed memory leak in sfRoute::generate() (closes #9886 - patch from acf0)
 * [32892] fixed Doctrine form generation when a schema has abstract tables (closes #9885 - patch from chok)
 * [32891] fixed sfFinder bug when mirroring file with twice the path inside it (closes #9892 - based on a patch from shordeaux)
 * [32890] fixed bad value of property 'list-style' in css of WebDebug toolbar (closes #9891 - patch from tomi)

08/05/11: Version 1.4.13
------------------------

 * [32707] fixed project:optimize which did not worked with plugins (closes #8760 - based on a patch from Pavel.Campr)
 * [32729] updated getHost() to consider a chain of forwarded hosts
 * [32740] fixed typo in error message (closes #9858)
 * [32741] removed non-countries from sfCultureInfo::getCountries() (closes #9848)
 * [32807] fixed parent for choice widgets (closes #8894)
 * [32808] simplified code from previous commit
 * [32810] fixed typo (closes #9866)
 * [32835] fixed previous commit when the renderer does not have the translate_choices option (see sfFormExtraPlugin for some examples)
 * [32836] fixed guessFromFileBinary in class sfvalidatorFile does not support MIME type including dot (closes #9871)
 * [32845] removed dead code (closes #9881)

07/01/11: Version 1.4.12
------------------------

 * [32634] fixed typo in an error message (closes #9762 - based on a patch from garak)
 * [32635] fixed typo (closes #9682)
 * [32637] added missing translations for fr (patch from Garfield-fr - closes #9025)
 * [32638] added missing translations for de (patch from _GeG_ - closes #9659)
 * [32639] fixed typo (closes #9561, patch from Elvis_the_King)
 * [32641] fixed extra code in stack trace (closes #9511, patch from tomi)
 * [32652] fixed sfRoute, which was not serializing some parameters (closes #9833)
 * [32653] fixed possible PHP notice (closes #9813)
 * [32654] fixed standard actions in sfObjectRouteCollection (HEAD is equivalent to GET, closes #9669, patch from ganchiku)
 * [32677] fixed missing information about Swiftmailer version (closes #9838)
 * [32678] fixed sfNumberFormat::format() when the value has a negative exponent (closes #9836)

03/30/11: Version 1.4.11
------------------------

 * Switched Doctrine external to our new mirror of the GitHub master branch

03/21/11: Version 1.4.10
------------------------

 * Updated to Doctrine 1.2.4 (security release)

02/04/11: Version 1.4.9
-----------------------

 * [31928] copied more values into a cachable partial's cloned response when the partial is marked as contextual in cache.yml (closes #7192)

lib/helper/TextHelper.php
<?php

/*
 * This file is part of the symfony package.
 * (c) 2004-2006 Fabien Potencier <fabien.potencier@symfony-project.com>
 * (c) 2004 David Heinemeier Hansson
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

/**
 * TextHelper.
 *
 * @package    symfony
 * @subpackage helper
 * @author     Fabien Potencier <fabien.potencier@symfony-project.com>
 * @author     David Heinemeier Hansson
 * @version    SVN: $Id: TextHelper.php 33022 2011-09-15 05:27:12Z fabien $
 */

/**
 * Truncates +text+ to the length of +length+ and replaces the last three characters with the +truncate_string+
 * if the +text+ is longer than +length+.
 */
function truncate_text($text, $length = 30, $truncate_string = '...', $truncate_lastspace = false)
{
  if ($text == '')
  {
    return '';
  }

  $mbstring = extension_loaded('mbstring');
  if($mbstring)
  {
   $old_encoding = mb_internal_encoding();
   @mb_internal_encoding(mb_detect_encoding($text));
  }
  $strlen = ($mbstring) ? 'mb_strlen' : 'strlen';
  $substr = ($mbstring) ? 'mb_substr' : 'substr';

  if ($strlen($text) > $length)
  {
    $truncate_text = $substr($text, 0, $length - $strlen($truncate_string));
    if ($truncate_lastspace)
    {
      $truncate_text = preg_replace('/\s+?(\S+)?$/', '', $truncate_text);
    }
    $text = $truncate_text.$truncate_string;
  }

  if($mbstring)
  {
   @mb_internal_encoding($old_encoding);
  }

  return $text;
}

/**
 * Highlights the +phrase+ where it is found in the +text+ by surrounding it like
 * <strong class="highlight">I'm a highlight phrase</strong>. The highlighter can be specialized by
 * passing +highlighter+ as single-quoted string with \1 where the phrase is supposed to be inserted.
 * N.B.: The +phrase+ is sanitized to include only letters, digits, and spaces before use.
 *
 * @param string $text subject input to preg_replace.
 * @param string $phrase string or array of words to highlight
 * @param string $highlighter regex replacement input to preg_replace.
 *
 * @return string
 */
function highlight_text($text, $phrase, $highlighter = '<strong class="highlight">\\1</strong>')
{
  if (empty($text))
  {
    return '';
  }

  if (empty($phrase))
  {
    return $text;
  }

  if (is_array($phrase) or ($phrase instanceof sfOutputEscaperArrayDecorator))
  {
    foreach ($phrase as $word)
    {
      $pattern[] = '/('.preg_quote($word, '/').')/i';
      $replacement[] = $highlighter;
    }
  }
  else
  {
    $pattern = '/('.preg_quote($phrase, '/').')/i';
    $replacement = $highlighter;
  }

  return preg_replace($pattern, $replacement, $text);
}

/**
 * Extracts an excerpt from the +text+ surrounding the +phrase+ with a number of characters on each side determined
 * by +radius+. If the phrase isn't found, nil is returned. Ex:
 *   excerpt("hello my world", "my", 3) => "...lo my wo..."
 * If +excerpt_space+ is true the text will only be truncated on whitespace, never inbetween words.
 * This might return a smaller radius than specified.
 *   excerpt("hello my world", "my", 3, "...", true) => "... my ..."
 */
function excerpt_text($text, $phrase, $radius = 100, $excerpt_string = '...', $excerpt_space = false)
{
  if ($text == '' || $phrase == '')
  {
    return '';
  }

  $mbstring = extension_loaded('mbstring');
  if($mbstring)
  {
    $old_encoding = mb_internal_encoding();
    @mb_internal_encoding(mb_detect_encoding($text));
  }
  $strlen = ($mbstring) ? 'mb_strlen' : 'strlen';
  $strpos = ($mbstring) ? 'mb_strpos' : 'strpos';
  $strtolower = ($mbstring) ? 'mb_strtolower' : 'strtolower';
  $substr = ($mbstring) ? 'mb_substr' : 'substr';

  $found_pos = $strpos($strtolower($text), $strtolower($phrase));
  $return_string = '';
  if ($found_pos !== false)
  {
    $start_pos = max($found_pos - $radius, 0);
    $end_pos = min($found_pos + $strlen($phrase) + $radius, $strlen($text));
    $excerpt = $substr($text, $start_pos, $end_pos - $start_pos);
    $prefix = ($start_pos > 0) ? $excerpt_string : '';
    $postfix = $end_pos < $strlen($text) ? $excerpt_string : '';

    if ($excerpt_space)
    {
      // only cut off at ends where $exceprt_string is added
      if($prefix)
      {
        $excerpt = preg_replace('/^(\S+)?\s+?/', ' ', $excerpt);
      }
      if($postfix)
      {
        $excerpt = preg_replace('/\s+?(\S+)?$/', ' ', $excerpt);
      }
    }

    $return_string = $prefix.$excerpt.$postfix;
  }

  if($mbstring)
  {
   @mb_internal_encoding($old_encoding);
  }
  return $return_string;
}

/**
 * Word wrap long lines to line_width.
 */
function wrap_text($text, $line_width = 80)
{
  return preg_replace('/(.{1,'.$line_width.'})(\s+|$)/s', "\\1\n", preg_replace("/\n/", "\n\n", $text));
}

/**
 * Returns +text+ transformed into html using very simple formatting rules
 * Surrounds paragraphs with <tt>&lt;p&gt;</tt> tags, and converts line breaks into <tt>&lt;br /&gt;</tt>
 * Two consecutive newlines(<tt>\n\n</tt>) are considered as a paragraph, one newline (<tt>\n</tt>) is
 * considered a linebreak, three or more consecutive newlines are turned into two newlines
 */
function simple_format_text($text, $options = array())
{
  $css = (isset($options['class'])) ? ' class="'.$options['class'].'"' : '';

  $text = sfToolkit::pregtr($text, array("/(\r\n|\r)/"        => "\n",               // lets make them newlines crossplatform
                                         "/\n{2,}/"           => "</p><p$css>"));    // turn two and more newlines into paragraph

  // turn single newline into <br/>
  $text = str_replace("\n", "\n<br />", $text);
  return '<p'.$css.'>'.$text.'</p>'; // wrap the first and last line in paragraphs before we're done
}

/**
 * Turns all urls and email addresses into clickable links. The +link+ parameter can limit what should be linked.
 * Options are :all (default), :email_addresses, and :urls.
 *
 * Example:
 *   auto_link("Go to http://www.symfony-project.com and say hello to fabien.potencier@example.com") =>
 *     Go to <a href="http://www.symfony-project.com">http://www.symfony-project.com</a> and
 *     say hello to <a href="mailto:fabien.potencier@example.com">fabien.potencier@example.com</a>
 */
function auto_link_text($text, $link = 'all', $href_options = array(), $truncate = false, $truncate_len = 35, $pad = '...')
{
  if ($link == 'all')
  {
    return _auto_link_urls(_auto_link_email_addresses($text), $href_options, $truncate, $truncate_len, $pad);
  }
  else if ($link == 'email_addresses')
  {
    return _auto_link_email_addresses($text);
  }
  else if ($link == 'urls')
  {
    return _auto_link_urls($text, $href_options, $truncate, $truncate_len, $pad);
  }
}

/**
 * Turns all links into words, like "<a href="something">else</a>" to "else".
 */
function strip_links_text($text)
{
  return preg_replace('/<a[^>]*>(.*?)<\/a>/s', '\\1', $text);
}

if (!defined('SF_AUTO_LINK_RE'))
{
  define('SF_AUTO_LINK_RE', '~
    (                       # leading text
      <\w+.*?>|             #   leading HTML tag, or
      [^=!:\'"/]|           #   leading punctuation, or
      ^                     #   beginning of line
    )
    (
      (?:https?://)|        # protocol spec, or
      (?:www\.)             # www.*
    )
    (
      [-\w]+                   # subdomain or domain
      (?:\.[-\w]+)*            # remaining subdomains or domain
      (?::\d+)?                # port
      (?:/(?:(?:[\~\w\+%-]|(?:[,.;:][^\s$]))+)?)* # path
      (?:\?[\w\+%&=.;-]+)?     # query string
      (?:\#[\w\-/\?!=]*)?        # trailing anchor
    )
    ([[:punct:]]|\s|<|$)    # trailing text
   ~x');
}

/**
 * Turns all urls into clickable links.
 */
function _auto_link_urls($text, $href_options = array(), $truncate = false, $truncate_len = 40, $pad = '...')
{
  $href_options = _tag_options($href_options);

  $fn = function ($matches) use ($href_options)
  {
    if (preg_match("/<a\s/i", $matches[1]))
    {
      return $matches[0];
    }
    else
    {
      return $matches[1].'<a href="'.($matches[2] == "www." ? "http://www." : $matches[2]).$matches[3].'"'.$href_options.'>'.$matches[2].$matches[3].'</a>'.$matches[4];
    }
  };

  $fnTruncated = function ($matches) use ($href_options, $truncate_len, $pad)
  {
    if (preg_match("/<a\s/i", $matches[1]))
    {
      return $matches[0];
    }
    else
    {
      if (strlen($matches[2].$matches[3]) > $truncate_len)
      {
        return $matches[1].
               '<a href="'.($matches[2] == "www." ? "http://www." : $matches[2]).$matches[3].'"'.$href_options.'>'.
               substr($matches[2].$matches[3], 0, $truncate_len).$pad.'</a>'.$matches[4];
      }

      else
      {
        return $matches[1].
               '<a href="'.($matches[2] == "www." ? "http://www." : $matches[2]).$matches[3].'"'.$href_options.'>'.
               $matches[2].$matches[3].'</a>'.$matches[4];
      }
    }
  };

  $callback_function = '
    if (preg_match("/<a\s/i", $matches[1]))
    {
      return $matches[0];
    }
    ';

  if ($truncate)
  {
    $callback_function .= '
      else if (strlen($matches[2].$matches[3]) > '.$truncate_len.')
      {
        return $matches[1].\'<a href="\'.($matches[2] == "www." ? "http://www." : $matches[2]).$matches[3].\'"'.$href_options.'>\'.substr($matches[2].$matches[3], 0, '.$truncate_len.').\''.$pad.'</a>\'.$matches[4];
      }
      ';
  }

  $callback_function .= '
    else
    {
      return $matches[1].\'<a href="\'.($matches[2] == "www." ? "http://www." : $matches[2]).$matches[3].\'"'.$href_options.'>\'.$matches[2].$matches[3].\'</a>\'.$matches[4];
    }
    ';

  return preg_replace_callback(
    SF_AUTO_LINK_RE,
    ($truncate ? $fnTruncated : $fn),
    $text
  );
}

/**
 * Turns all email addresses into clickable links.
 */
function _auto_link_email_addresses($text)
{
  // Taken from http://snippets.dzone.com/posts/show/6156
  return preg_replace("#(^|[\n ])([a-z0-9&\-_\.]+?)@([\w\-]+\.([\w\-\.]+\.)*[\w]+)#i", "\\1<a href=\"mailto:\\2@\\3\">\\2@\\3</a>", $text);

  // Removed since it destroys already linked emails 
  // Example:   <a href="mailto:me@example.com">bar</a> gets <a href="mailto:me@example.com">bar</a> gets <a href="mailto:<a href="mailto:me@example.com">bar</a>
  //return preg_replace('/([\w\.!#\$%\-+.]+@[A-Za-z0-9\-]+(\.[A-Za-z0-9\-]+)+)/', '<a href="mailto:\\1">\\1</a>', $text);
}
